### PR TITLE
feat: add blog about property-based testing w python ml

### PR DESCRIPTION
**Pull request by stijnmoreels**
Adds a new blog post on the relationship of property-based testing and machine learning with Python, all discovered when maintaining the Arcus ML repo.

Relates to #187

### CODE REVIEW

File: community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
+# Property-based testing the Arcus Python implementations for streamlined ML development
+Arcus.ML contains several easy-to-use Python implementations for more fluent ML development. This post will guide through the process of testing these implementations with a technique to ensure robust outcomes: property-based testing.
```
Comment (pim-simons) [inline community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md:2]:
```suggestion
Arcus.ML contains several easy-to-use Python implementations for more fluent machine learning development. This post will guide through the process of testing these implementations with a technique to ensure robust outcomes: property-based testing.
```

Maybe better to not use an abbreviation?

File: community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+import arcus.ml.images.conversion as conv
+
+@given(images())
+def test_black_white_property(image_list):
+    image_converted = conv.to_blackwhite(image_list)
+    assert all_black_and_white(image_converted)
+```
+
+## Customizable time windows for more accurate predication
+A big portion of the Arcus.ML library is about manipulating time series: combining, slicing, time references... For this post we'll be looking at the windows function which will provide us with a 'windowed' version of an incoming time series. Looking at time series in windows helps greatly in a more accurate prediction at a given point with the recent past. Put simply, windows are nothing more than a set of smaller versions of the initial data set, each starting from a different place and overlapping with each other. Time series are expressed here as [pandas](https://pandas.pydata.org/docs/getting_started/index.html#getting-started) data frames and will output the windows as [numpy](https://numpy.org/doc/stable/user/absolute_beginners.html) arrays.
```
Comment (pim-simons) [inline community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md:67]:
```suggestion
A big portion of the Arcus.ML library is about manipulating time series: combining, slicing, time references... For this post we'll be looking at the windows function which will provide us with a 'windowed' version of an incoming time series. Looking at time series in windows helps greatly to get a more accurate prediction for a given point with the recent past. Put simply, windows are nothing more than a set of smaller versions of the initial data set, each starting from a different place and overlapping with each other. Time series are expressed here as [pandas](https://pandas.pydata.org/docs/getting_started/index.html#getting-started) data frames and will output the windows as [numpy](https://numpy.org/doc/stable/user/absolute_beginners.html) arrays. 
```

File: community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    win_size = 6
+    # Act
+    windows, _ = tops.get_windows(df, win_size)
+    # Assert
+    col_size = len(df.columns)
+    assert windows.shape == (15, win_size, col_size)
+    assert_all_windows_equals(df, windows, win_size)
+```
+
+Generating a table of sensor data as data frames is not so hard as one might think. Hypothesis already provides a way to [generate data frames](https://hypothesis.readthedocs.io/en/latest/numpy.html?highlight=dataframe#hypothesis.extra.pandas.data_frames) for us. All we have to do is provide it with some reasonable size and types. In our first version, we used a subset of our initial sensor data columns and allowed Hypothesis to generate for us any non-Nan float number.
```
Comment (pim-simons) [inline community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md:111]:
```suggestion
Generating a table of sensor data as data frames is not as hard as one might think. Hypothesis already provides a way to [generate data frames](https://hypothesis.readthedocs.io/en/latest/numpy.html?highlight=dataframe#hypothesis.extra.pandas.data_frames) for us. All we have to do is provide it with some reasonable size and types. In our first version, we used a subset of our initial sensor data columns and allowed Hypothesis to generate for us any non-Nan float number.
```

**Review by pim-simons — APPROVED** [review]
Great read, understandable even for a machine learning noob as me 😉 
Only reviewed the text as I don't have the knowledge to actually check if the python stuff is correct.

**Comment by stijnmoreels** [comment]
> Great read, understandable even for a machine learning noob as me 😉 Only reviewed the text as I don't have the knowledge to actually check if the python stuff is correct.

Hihi, 😄 , you and me both, mate. Bit a rookie myself in this topic. The code samples are actually a copy-paste as how we do it in Arcus ML, so the workings are already verified. It was also reviewed by Alexis, so I think we can say that it passed the test there too 😉 . I think I also mentioned that this post and this idea is work-in-progress and by no means the end-result.

File: community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md (hunk 1 above)
Comment (stijnmoreels) [reply community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md:2]: 👍 Yes, is good!

**Comment by stijnmoreels** [comment]
@ALaks96 @fgheysels any comments here?

File: community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+from hypothesis import given
+import arcus.ml.images.conversion as conv
+
+@given(images())
+def test_black_white_property(image_list):
+    image_converted = conv.to_blackwhite(image_list)
+    assert all_black_and_white(image_converted)
+```
+
+## Customizable time windows for more accurate predication
```
Comment (fgheysels) [inline community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md:66]: predication or prediction ?

**Review by fgheysels — APPROVED** [review]
I don't know anything about ML but it is well-written!

**Comment by stijnmoreels** [comment]
> I don't know anything about ML but it is well-written!

Luckily, there isn't much ML stuff going on as the blog is mostly about property-based testing. It's also written in the post that this is by all means no definite solution but more a discovery. Let's hope, when we finally get some ML experts on Arcus, that we can write more specific posts on this topic.

File: community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md (hunk 4 above)
Comment (stijnmoreels) [reply community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md:66]: prediction 😅 . Don't need to preach here. 😄

**Pull request merged**